### PR TITLE
Custom Scripts in IncidentFields and Layouts

### PR DESCRIPTION
**Pull request by merit-maita**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/XSUP-18812

## Description
fixed an issue where custom scripts names were replaced with uuids.

### CODE REVIEW

File: demisto_sdk/commands/download/downloader.py [hunk 1]
```diff
@@ -223,9 +223,10 @@ def download_playbook_yaml(self, playbook_string) -> str:
         playbook_id = file_yaml_object.get('id')
         playbook_name = file_yaml_object.get('name')
         if playbook_id and playbook_name and (playbook_name in self.input_files or self.all_custom_content):
-            # download the playbook yaml in case playbook name appears in the input_files or --all-custom-content flag is true
+            # download the playbook yaml in case playbook name appears in the input_files
+            # or --all-custom-content flag is true
 
-            # this will make sure that we save the downloaded files in the custom cotent temp dir
+            # this will make sure that we save the downloaded files in the custom content temp dir
```
Comment (DeanArbel) [inline demisto_sdk/commands/download/downloader.py:229]: nice

File: demisto_sdk/commands/download/downloader.py [hunk 2]
```diff
@@ -236,6 +237,63 @@ def download_playbook_yaml(self, playbook_string) -> str:
...
+                            layout_string = layout_string.replace(item.get('scriptId'),
+                                                                  scripts_mapper[item.get('scriptId')])
+
+        return layout_string
+
+    @staticmethod
+    def map_script(script_string: str, scripts_mapper: dict) -> dict:
+        script_yml = yaml.load(script_string)
+        script_id = script_yml.get('commonfields').get('id')
+        if re.search(r'^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$',
```
Comment (DeanArbel) [inline demisto_sdk/commands/download/downloader.py:272]: Please extract the regex expression to the `constants.py` file, ideally reusing the other existing `constants` to construct the `regex`.

File: demisto_sdk/commands/download/downloader.py [hunk 3]
```diff
@@ -236,6 +237,63 @@ def download_playbook_yaml(self, playbook_string) -> str:
...
+                     script_id):
+            scripts_mapper[script_id] = script_yml.get('name')
+        return scripts_mapper
+
+    def handle_file(self, string_to_write, member_name, scripts_id_name):
+
+        if 'automation-' in member_name:
+            scripts_id_name = self.map_script(string_to_write, scripts_id_name)
+
+        if not self.list_files and re.search(r'incidentfield-.*\.json', member_name):
```
Comment (DeanArbel) [inline demisto_sdk/commands/download/downloader.py:282]: Same as above.

File: demisto_sdk/commands/download/downloader.py [hunk 4]
```diff
@@ -236,6 +237,63 @@ def download_playbook_yaml(self, playbook_string) -> str:
...
+            string_to_write = self.handle_incidentfield(string_to_write, scripts_id_name)
+
+        if not self.list_files and re.search(r'playbook-.*\.yml', member_name):
```
Comment (DeanArbel) [inline demisto_sdk/commands/download/downloader.py:285]: Same as above.

File: demisto_sdk/commands/download/downloader.py [hunk 5]
```diff
@@ -236,6 +237,63 @@ def download_playbook_yaml(self, playbook_string) -> str:
...
+            #  if the content item is playbook and list-file flag is true, we should download the
+            #  file via direct REST API because there are props like scriptName, that playbook from custom
+            #  content bundle don't contain
+
+            string_to_write = self.download_playbook_yaml(string_to_write)
+
+        if not self.list_files and re.search(r'layout.*\.json', member_name):
```
Comment (DeanArbel) [inline demisto_sdk/commands/download/downloader.py:292]: Same as above.

**Review by DeanArbel — CHANGES_REQUESTED** [review]
Looks good, just make sure to extract the content items identifiers to constants, and tag @dorschw as a reviewer once you're done.

**Review by DeanArbel — APPROVED** [review]
Good job!

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,7 @@
 * Added hidden properties to integration command argument and script argument.
 * Added `--override-existing` to **upload** that skips the confirmation prompt for overriding existing content packs. @mattbibbydw
 * Fixed an issue where **validate** failed in private repo due to attempts to read from nonexisting file.
+* Fixed an issue where custom scripts names were replaced with uuid in IncidentFields and Layouts.
```
Comment (dorschw) [inline CHANGELOG.md:7]:
```suggestion
* Fixed an issue in **download**, where names of custom scripts were replaced with UUIDs in IncidentFields and Layouts.
```

File: demisto_sdk/commands/download/downloader.py [hunk 7]
```diff
@@ -236,6 +238,62 @@ def download_playbook_yaml(self, playbook_string) -> str:
...
         return playbook_string
 
+    def handle_incidentfield(self, incidentfield_string: str, scripts_mapper: dict) -> str:
+        # In case the incident field uses a custom script, replace the id value of the script with its name
+        file_json_object = json.loads(incidentfield_string)
+        incidentfield_name = file_json_object.get('name')
+        if incidentfield_name and (incidentfield_name in self.input_files or self.all_custom_content):
+            if file_json_object.get('script') and file_json_object.get('script') in scripts_mapper:
+                incidentfield_string = incidentfield_string.replace(file_json_object.get('script'),
+                                                                    scripts_mapper[file_json_object.get('script')])
```
Comment (dorschw) [inline demisto_sdk/commands/download/downloader.py:248]: please replace the multiple `.get('script')`s with a single call saving its value as a variable.

File: demisto_sdk/commands/download/downloader.py [hunk 8]
```diff
@@ -236,6 +238,62 @@ def download_playbook_yaml(self, playbook_string) -> str:
...
+        # In case the layout uses a custom script, replace the id value of the script with its name
+        file_json_object = json.loads(layout_string)
+        layout_name = file_json_object.get('name')
+        if layout_name and (layout_name in self.input_files or self.all_custom_content):
+            tabs = file_json_object.get('detailsV2', {}).get('tabs', [])
+            for tab in tabs:
+                sections = tab.get('sections', [])
+                for section in sections:
+                    items = section.get('items', [])
+                    for item in items:
```
Comment (dorschw) [inline demisto_sdk/commands/download/downloader.py:262]:
```suggestion
            for tab in file_json_object.get('detailsV2', {}).get('tabs', ()):
                for section in tab.get('sections', ()):
                    for item in section.get('items', ()):
```

File: demisto_sdk/commands/download/downloader.py [hunk 9]
```diff
@@ -236,6 +238,62 @@ def download_playbook_yaml(self, playbook_string) -> str:
...
+                        if item.get('scriptId') and item.get('scriptId') in scripts_mapper:
+                            layout_string = layout_string.replace(item.get('scriptId'),
+                                                                  scripts_mapper[item.get('scriptId')])
```
Comment (dorschw) [inline demisto_sdk/commands/download/downloader.py:265]: please refrain from repetitive `.get` calls - use a variable instead.

File: demisto_sdk/commands/common/constants.py [hunk 10]
```diff
@@ -593,6 +593,10 @@ class FileType(str, Enum):
 
 INDICATOR_TYPES_REPUTATIONS_REGEX = fr'{CAN_START_WITH_DOT_SLASH}{INDICATOR_TYPES_DIR}.reputations\.json$'
 
+INCIDENT_FIELD_REGEX = r'incidentfield-.*\.json'
+
+LAYOUT_REGEX = r'layout.*\.json'
```
Comment (dorschw) [inline demisto_sdk/commands/common/constants.py:598]:
```suggestion
INCIDENT_FIELD_FILE_NAME_REGEX = r'incidentfield-.*\.json'

LAYOUT_FILE_NAME__REGEX = r'layout.*\.json'
```

**Review by dorschw — COMMENTED** [review]
Nicely done

**Pull request merged**